sqlair.go

```diff
@@ -87,64 +89,122 @@ func (db *DB) Query(ctx context.Context, s *Statement, inputArgs ...any) *Query
 	}
 
 	qe, err := s.pe.Query(inputArgs...)
-	return &Query{qs: db.db, qe: qe, err: err, ctx: ctx}
+	return &Query{qs: db.db, qe: qe, ctx: ctx, err: err}
 }
 
-// Run will execute the query.
-// Any rows returned by the query are ignored.
+// Run is an alias for Get that takes no arguments.
 func (q *Query) Run() error {
+	return q.Get()
+}
+
+// Query.Get will run the query.
+// The first row (if there is one) will be scanned into the outputArgs.
+// It will give ErrNoRows if the query contains output expressions and there are no rows.
+// A pointer to an Outcome struct can be passed as the first argument which will be populated with the outcome of the query.
+func (q *Query) Get(outputArgs ...any) error {
 	if q.err != nil {
 		return q.err
 	}
-	_, err := q.qs.ExecContext(q.ctx, q.qe.QuerySQL(), q.qe.QueryArgs()...)
-	if err != nil {
-		return err
+	var outcome *Outcome
+	if len(outputArgs) > 0 {
+		if oc, ok := outputArgs[0].(*Outcome); ok {
+			outcome = oc
+			outputArgs = outputArgs[1:]
+		}
 	}
-	return nil
+
+	iter := q.Iter()
+	defer iter.Close()
+	if outcome != nil {
+		err := iter.Get(outcome)
+		if err != nil {
+			return err
+		}
+	}
+	if !q.qe.HasOutputs() {
+		if len(outputArgs) > 0 {
+			return fmt.Errorf("cannot get results: query does not return any results")
```

Comment:
This is the case where they have run a DML statement like "INSERT" or "UPDATE" and they have asked for outputs. We don't actually call `Get` (apart from before `Next` to get the `Outcome`) because `Next` returns false. So there is no case where `Get` is called with these `outputArgs`.

This is a mistake from the user where their query contains no output expressions but they have passed some structs to the results into.

On further consideration I think this error should be moved to before the query is actually executed, since if the error is being triggered the user has misunderstood something.

Reply:
The question was actually whether this error would be picked up by the expression parser itself. It should be, I think: the expression says there are not outputs, and yet we've passed data types in that are missing from the expression.

Reply:
True, the error message is a bit unclear here, I'll change it.